firedrake/adjoint/ensemble_reduced_functional.py

```diff
@@ -53,15 +61,72 @@ class EnsembleReducedFunctional(ReducedFunctional):
     works, please refer to the `Firedrake manual
     <https://www.firedrakeproject.org/parallelism.html#id8>`_.
     """
-    def __init__(self, J, control, ensemble, scatter_control=True):
-        super(EnsembleReducedFunctional, self).__init__(J, control)
+    def __init__(self, J, control, ensemble, scatter_control=True,
+                 gather_functional=None):
         self.ensemble = ensemble
+        if isinstance(J, list):
+            assert isinstance(control, list), "Controls should also be a list."
+            assert len(control) % len(J) == 0, "Controls and J have mismatching lengths."
```

Comment:
Instead of using `assert`, use exception raising errors:

```
if not isinstance(control, list):
    raise TypeError("Controls should be a list.")
if len(control) % len(J) != 0:
    raise ValueError("Controls and J have mismatching lengths.")
```